spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -109,29 +110,96 @@ public GqlDatastoreQuery(Class<T> type, DatastoreQueryMethod queryMethod,
 		this.gql = StringUtils.trimTrailingCharacter(gql.trim(), ';');
 	}
 
+	private static Object getNonEntityObjectFromRow(Object x) {
+		Object mappedResult;
+		if (x instanceof Key) {
+			mappedResult = x;
+		}
+		else {
+			BaseEntity entity = (BaseEntity) x;
+			Set<String> colNames = entity.getNames();
+			if (colNames.size() > 1) {
+				throw new DatastoreDataException(
+						"The query method returns non-entity types, but the query result has "
+								+ "more than one column. Use a Projection entity type instead.");
+			}
+			mappedResult = entity.getValue((String) colNames.toArray()[0]).get();
+		}
+		return mappedResult;
+	}
+
 	@Override
 	public Object execute(Object[] parameters) {
 		GqlQuery query = bindArgsToGqlQuery(this.gql, getParamTags(), parameters);
-		Iterable found = this.datastoreTemplate.queryKeysOrEntities(query, this.entityType);
+		boolean returnTypeIsCollection = this.queryMethod.isCollectionQuery();
+		Class returnedItemType = this.queryMethod.getReturnedObjectType();
+
+		boolean isNonEntityReturnType = isNonEntityReturnedType(returnedItemType);
+
+		Iterable found = isNonEntityReturnType
+				? this.datastoreTemplate.query(query,
+						GqlDatastoreQuery::getNonEntityObjectFromRow)
+				: this.datastoreTemplate.queryKeysOrEntities(query, this.entityType);
+
 		List rawResult = found == null ? Collections.emptyList()
 				: (List) StreamSupport.stream(found.spliterator(), false)
 						.collect(Collectors.toList());
+
 		Object result;
+
 		if (this.queryMethod.isCountQuery()) {
 			result = rawResult.size();
 		}
 		else if (this.queryMethod.isExistsQuery()) {
 			result = !rawResult.isEmpty();
 		}
-		else if (rawResult.isEmpty() || rawResult.get(0).getClass() == Key.class) {
-			result = rawResult;
+		else if (!returnTypeIsCollection) {
+			if (rawResult.isEmpty()) {
```

Comment:
move both `rawResult.isEmpty()` and `rawResult.size() > 1` checks into `convertSingularResult` method